### PR TITLE
JENKINS-26828 Add support for installing the NodeJS windows msi

### PR DESCRIPTION
**Pull request by iaincoulter**
Fix for  installing msi on windows.

also fixes issues with JENKINS-21953 NodeJs plugin doesn't work on windows based Jenkins machine

### CODE REVIEW

**Comment by fcamblor** [comment]
Great !

It seems to be a nice addition. I just need some time to for non-regression testing, and if everything is OK, I'll merge it and make a new release of the plugin.

**Comment by iaincoulter** [comment]
is everything ok for doing a merge?

**Comment by destroyerofbuilds** [comment]
Also resolves the following: https://issues.jenkins-ci.org/browse/JENKINS-26828

**Comment by destroyerofbuilds** [comment]
Sorry for the earlier post.

I too would like to see Windows support for installing Node binaries within a build environment.

File: src/main/java/jenkins/plugins/nodejs/tools/NodeJSInstaller.java [hunk 1]
```diff
@@ -129,19 +132,36 @@ public FilePath performInstallation(ToolInstallation tool, Node node, TaskListen
...
+            }
+            if(result) {
                 expected.child(".timestamp").delete(); // we don't use the timestamp
-                String archiveIntermediateDirectoryName = installerPathResolver.extractArchiveIntermediateDirectoryName(relativeDownloadPath);
+                
+                String archiveIntermediateDirectoryName = (platform == NodeJSInstaller.Platform.WINDOWS) ? 
+                            "nodejs" : installerPathResolver.extractArchiveIntermediateDirectoryName(relativeDownloadPath);
                 this.pullUpDirectory(expected, archiveIntermediateDirectoryName);
-
+
```
Comment (destroyerofbuilds) [inline src/main/java/jenkins/plugins/nodejs/tools/NodeJSInstaller.java]: Can you remove this extra whitespace?

File: src/main/java/jenkins/plugins/nodejs/tools/NodeJSInstaller.java [hunk 2]
```diff
@@ -149,7 +169,12 @@ public FilePath performInstallation(ToolInstallation tool, Node node, TaskListen
                 expected.child(NPM_PACKAGES_RECORD_FILENAME).delete();
                 ArgumentListBuilder npmScriptArgs = new ArgumentListBuilder();
 
-                FilePath npmExe = expected.child("bin/npm");
+                FilePath npmExe = expected.child(platform ==  Platform.WINDOWS ? "npm.cmd" : "bin/npm");
+                if (platform ==  Platform.WINDOWS)
```
Comment (destroyerofbuilds) [inline src/main/java/jenkins/plugins/nodejs/tools/NodeJSInstaller.java]: You have an extra space after `==`.

**Comment by iamcdonald** [comment]
Are there any plans to merge this in and release it any time soon?

**Comment by dabperceptive** [comment]
:+1:

**Comment by kmthinkingrock** [comment]
+1 for merging this ...

**Comment by fbauzac** [comment]
+1: indeed supporting the installation of the official .msi file would remove the burden to have to extract the official .msi somewhere, zipping the files back into a .zip and store them into a repository...

**Comment by nfalco79** [comment]
@iamcdonald I have test your good PR.
Actually it does not work because the constructed URL https://nodejs.org/dist/v5.12.0/**x64/**node-v5.12.0-x64.msi is wrong (remove the bold part to be working) for x64 machine.

To be able to merge please:
- squash your commits to have a clean history and to remove revert commits
- make code formatting coherent with actual code and your new code (for example braces sometimes are on same line sometimes not)

**Comment by iamcdonald** [comment]
@nfalco79 I think you meant to tag @iaincoulter. 
It seems to be his PR rather than mine.

Thanks.

**Comment by iaincoulter** [comment]
Is my code even needed anymore ? 

Originally i was working against 0.9.9 hence the URL was fine. 

But if you look at the latest releases https://nodejs.org/dist/v7.2.0/ 

Not only is a msi provided but so is a 7z and zip file ? which should work a lot more similarly to how linux works with the .tar.gz 
node-v7.2.0-win-x64.7z 22-Nov-2016 23:01             7500031 node-v7.2.0-win-x64.zip 22-Nov-2016 23:01            13171366 node-v7.2.0-win-x86.7z 22-Nov-2016 22:58             6456848 node-v7.2.0-win-x86.zip 22-Nov-2016 22:58            11762677 node-v7.2.0-x64.msi 22-Nov-2016 23:01            13762560 node-v7.2.0-x86.msi 22-Nov-2016 22:58            12255232 
-- 
Iain

**Comment by nfalco79** [comment]
@iamcdonald sorry my mistake due autocompletition.

@iaincoulter ok I've not see in depth 0.x version. So download URL for windows has different path depending on version (very annoing) and not all version have archive available. The pluggable strategy must use also version.

**Comment by nfalco79** [comment]
@iaincoulter Do you plan to rework this PR?

**Comment by iaincoulter** [comment]
@nfalco79 At the moment have very little free time to be able to spend time working on this.

**Comment by nfalco79** [comment]
Ok I will continue the fix starting from your changes

**Comment by nfalco79** [comment]
I have complete the task in [this](https://github.com/jenkinsci/nodejs-plugin/tree/feature/JENKINS-26828) branch.

Maybe I have to rewrite 2 classes (Version and VersionRange) because are under APL2 license :(

**Pull request closed without merging**